File: nodejs/src/handlers/tls.js
'use strict';

function resolveProtocolByPort(port) {
  switch (port) {
    case 8442: return 'TLSv1.2';
    case 8443: return 'TLSv1.3';
    case 8444: return 'TLSv1.2 or TLSv1.3';
    default: return 'unknown';
  }
}

function tlsInfo(req, res) {
  const port = req.socket.localPort || parseInt(req.get('host')?.split(':')[1], 10) || 8080;
  const info = {
    timestamp: new Date().toISOString(),
    scheme: req.secure || req.protocol === 'https' ? 'https' : 'http',
    serverPort: port,
    remoteAddr: req.ip || req.socket.remoteAddress,
  };

  if (req.secure || req.protocol === 'https') {
    info.tlsEnabled = true;
    const proto = req.socket.getProtocol?.();
    if (proto) {
      info.negotiatedProtocol = proto;
      info.cipherSuite = req.socket.getCipher?.()?.name;
    } else {
      info.negotiatedProtocol = resolveProtocolByPort(port);
      info.note = 'SSLSession attribute unavailable; port-based protocol inferred';
    }
  } else {
    info.tlsEnabled = false;
    info.negotiatedProtocol = 'none (plain HTTP)';
    info.note = 'Connect via https:// to test TLS';
  }

  res.json(info);
}

module.exports = { tlsInfo };

File: go/internal/handler/tls.go
package handler

import (
	"crypto/tls"
	"encoding/json"
	"net/http"
	"strconv"
	"time"
)

func resolveProtocolByPort(port int) string {
	switch port {
	case 8442:
		return "TLSv1.2"
	case 8443:
		return "TLSv1.3"
	case 8444:
		return "TLSv1.2 or TLSv1.3"
	default:
		return "unknown"
	}
}

func TlsInfo(w http.ResponseWriter, r *http.Request) {
	port, _ := strconv.Atoi(r.URL.Port())
	if port == 0 {
		if r.TLS != nil {
			port = 443
		} else {
			port = 8080
		}
	}

	info := map[string]interface{}{
		"timestamp":  time.Now().UTC().Format(time.RFC3339Nano),
		"scheme":     "https",
		"serverPort": port,
		"remoteAddr": r.RemoteAddr,
	}

	if r.TLS != nil {
		info["tlsEnabled"] = true
		info["negotiatedProtocol"] = tlsVersionName(r.TLS.Version)
		info["cipherSuite"] = tls.CipherSuiteName(r.TLS.CipherSuite)
	} else if r.URL.Scheme == "https" || r.Header.Get("X-Forwarded-Proto") == "https" {
		info["tlsEnabled"] = true
		info["negotiatedProtocol"] = resolveProtocolByPort(port)
		info["note"] = "SSLSession attribute unavailable; port-based protocol inferred"
	} else {
		info["scheme"] = "http"
		info["tlsEnabled"] = false
		info["negotiatedProtocol"] = "none (plain HTTP)"
		info["note"] = "Connect via https:// to test TLS"
	}

	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(info)
}

func tlsVersionName(v uint16) string {
	switch v {
	case tls.VersionTLS13:
		return "TLSv1.3"
	case tls.VersionTLS12:
		return "TLSv1.2"
	default:
		return "unknown"
	}
}
